src/library_manager/lib_manager.c

```diff
@@ -0,0 +1,681 @@
+#include <sof/audio/component_ext.h>
+#include <sof/common.h>
+#include <sof/ipc/topology.h>
+#include <sof/lib/wait.h>
+
+#include <sof/sof.h>
+#include <sof/spinlock.h>
+#include <sof/lib_manager.h>
+#include <sof/audio/module_adapter/module/generic.h>
+#include <sof/audio/module_adapter/module/iadk_modules.h>
+
+#include <zephyr/drivers/mm/system_mm.h>
+
+#include <errno.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+/* 54cf5598-8b29-11ec-a8a3-0242ac120002 */
+
+DECLARE_SOF_UUID("lib_manager", lib_manager_uuid, 0x54cf5598, 0x8b29, 0x11ec,
+		 0xa8, 0xa3, 0x02, 0x42, 0xac, 0x12, 0x00, 0x02);
+
+DECLARE_TR_CTX(lib_manager_tr, SOF_UUID(lib_manager_uuid), LOG_LEVEL_INFO);
+
+/* TODO: Manifest size should be defined somewhere - to check!!!*/
+#define CAVS18_MAX_MANIFEST_SIZE  (38 * 1024)
+#define DMA_ELEM_SIZE		32
+#define PAGE_ALIGN			__aligned(CONFIG_MM_DRV_PAGE_SIZE)
+
+#define DMA_LOAD_BUFFER_SIZE   CAVS18_MAX_MANIFEST_SIZE
+
+/**
+ * DMA buffer
+ */
+struct lib_manager_dma_buf {
+	uintptr_t w_ptr;	/**< write pointer */
+	uintptr_t r_ptr;	/**< read pointer */
+	uintptr_t addr;		/**< buffer start pointer */
+	uintptr_t end_addr;	/**< buffer end pointer */
+	uint32_t size;		/**< buffer size */
+	uint32_t avail;		/**< buffer avail data */
+};
+
+struct lib_manager_dma_ext {
+	struct dma *dma;
+	struct dma_chan_data *chan;
+	struct lib_manager_dma_buf dmabp;
+};
+
+struct ext_library loader_ext_lib;
+
+void *lib_manager_get_config(struct comp_dev *dev)
+{
+	struct processing_module *mod = comp_get_drvdata(dev);
+	struct module_data *md = &mod->priv;
+	struct module_config *dst = &md->cfg;
+
+	return dst->data;
+}
+
+static int lib_manager_load_data_from_storage(uint8_t *vma, uint8_t *s_addr,
+					      uint32_t size, uint32_t flags)
+{
+	int ret = sys_mm_drv_map_region(vma, POINTER_TO_UINT(NULL),
+					size, SYS_MM_MEM_PERM_RW | SYS_MM_MEM_PERM_EXEC);
+	if (ret < 0) {
+		ret = -ENOMEM;
+		return ret;
+	}
+
+	ret = memcpy_s((void *)vma, size,
+		       (void *)s_addr, size);
+	if (ret < 0)
+		return ret;
+	dcache_writeback_region((void *)vma, size);
+
+	/* TODO: Change attributes for memory to FLAGS  */
+	return 0;
+}
+
+static int lib_manager_load_module(uint32_t module_id, struct sof_man_module *mod,
+				   struct sof_man_fw_desc *desc)
+{
+	struct ext_library *ext_lib = ext_lib_get();
+	uint32_t lib_id = LIB_MANAGER_GET_LIB_ID(module_id);
+	size_t load_offset = (size_t)((void *)ext_lib->desc[lib_id]);
+	void *va_base_text = (void *)mod->segment[SOF_MAN_SEGMENT_TEXT].v_base_addr;
+	void *src_txt = (void *)(mod->segment[SOF_MAN_SEGMENT_TEXT].file_offset + load_offset);
+	size_t st_text_size = mod->segment[SOF_MAN_SEGMENT_TEXT].flags.r.length;
+	void *va_base_rodata = (void *)mod->segment[SOF_MAN_SEGMENT_RODATA].v_base_addr;
+	void *src_rodata =
+		(void *)(mod->segment[SOF_MAN_SEGMENT_RODATA].file_offset + load_offset);
+	size_t st_rodata_size = mod->segment[SOF_MAN_SEGMENT_RODATA].flags.r.length;
```

Comment:
These should all be treated as untrusted and validated before use.

Reply:
Authentication, which is optional will be delivered in separate PR.

Reply:
sorry, this is not authentication - but a basic security validation that the inputs are all in allowed ranges. i.e. module is not attempting to overwrite base FW text with it's own text by using an incorrect v_base_addr.

Reply:
I understand that this protection is done by memory manager in function **lib_manager_load_data_from_storage()**. The memory manager API function **sys_mm_drv_map_region()** tries to map virtual memory. If this memory is already used (mapped), the function will fail and we will not be able to use it.